src/Artifacts.jl

```diff
@@ -0,0 +1,589 @@
+module Artifacts
+
+import Base: get, SHA1
+import ..depots1
+import ..GitTools: tree_hash, set_readonly
+using ..BinaryPlatforms
+import ..TOML
+import ..Types: parse_toml
+import ..PlatformEngines: download_verify_unpack, probe_platform_engines!, package
+
+export create_artifact, artifact_exists, artifact_path, remove_artifact, verify_artifact,
+       artifact_meta, artifact_hash, bind_artifact, unbind_artifact, download_artifact,
+       find_artifact_toml, ensure_artifact_installed, @artifact_str, archive_artifact
+
+## Philosophy of Artifacts:
+#
+# - At the lowest level, artifacts are content-addressed chunks of data.  Similarly to
+#   how packages are identified by UUIDs but have human-readable mappings through the
+#   registry, artifacts are identified by their SHA1 git tree hashes, but have human-
+#   readable mappings through `Artifact.toml` files.
+#
+# - Artifacts do not _need_ to have names bound to them; they are first and foremost
+#   identified by their tree hash.  A name is merely a pointer to a hash, and the hash
+#   itself is what is used in all operations except the initial lookup.
+#
+# - A single name can be associated with multiple hashes, where one is chosen at runtime.
+#   The canonical example of this is platform-dependent artifacts.  At `Artifact.toml`
+#   parse-time one of the list of hashes will be chosen as the best match for the
+#   current environment (Julia version, OS, processor architecture, etc...), and note
+#   that it is acceptable for there to be multiple possible choices (e.g. an AVX2-
+#   optimized build and a generic x86_64 build that will both run on a modern processor),
+#   however one single hash will be stably chosen from the multiple possibilities.
+#
+# - Artifacts can be created on-machine through the `create()` functional API, and can
+#   also be downloaded as tarballs from external webservers by embedding information into
+#   the `Artifact.toml` file.  This can be done programmatically through the
+#   `bind_artifact()` function.
+
+
+# Levels of API:
+#
+# - Base level:
+#   - create_artifact(): given no information, simply creates an artifact through a user-provided
+#     callback and returns the hash of the resultant file.
+#
+# - Hash level:
+#   - artifact_exists(hash): returns true if the hash exists on disk, false otherwise
+#   - artifact_path(hash):   returns the path to a hash on disk, throws if it does not exist
+#   - remove_artifact(hash): deletes `path(hash)` if it exists, does nothing otherwise.
```

Comment:
Is this a user facing API? Why is that needed and how can it be safe if e.g. other packages uses the same artifact?

Reply:
You're right; we shouldn't encourage users to call this; we should leave it up to `gc` in general.